
# Packages instalaltion

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("Biostrings", force = T)
BiocManager::install("Rsubread")
install.packages("fastqcr")
BiocManager::install("Rqc")
BiocManager::install("Rfastp")
```

# Packages loading

```{r}
#library('GenomicAlignments')
#library('Biostrings')
```

# Fastq files quality check

```{r}
library(Rqc)
rqc(path = "fastq/", # path to fastq files
    pattern = ".fastq.gz",  # pattern of analized files
    outdir = "fastqc/", # folder to save the report
    file = "rqc_kurs") # name of html report file
```

```{r}
library(Rfastp)
rfastp("fastp_test.fastq", # file to analysis
       outputFastq = "fastp_trimmed/fastp_test_trimmed.fastq") # path and name of new fastq file after filtering and adapters trimming and html report
```

```{r}
library("fastqcr") #CRAN
fastqc_install() # install only at first time
fastq.dir <- "/path/to/folder/fastq"

fastqc(fastq.dir, threads = 4) # 'threds' can be changed, depending on the computer
```

```{r}
qc.path <- "/path/to/fastq/folder/FASTQC"
qc <- qc_aggregate(qc.path, progressbar = T)
```

```{r}
head(qc)
summary(qc)
qc_stats(qc)
qc_fails(qc, "module") #
qc_problems(qc, "sample") # Samples with issues
```

# Mapping

```{r}
library(Rsubread)
```

## Genome index building

```{r}
# mouse genome (mm10), chromosome 19 is in file chr19.fa.gz'
# next command creates index files with names 'chr19_mm10'

buildindex(basename = "index_subread/chr19_mm10", reference = "ref/chr19.fa")

dir("index_subread/", pattern = "chr19") # show files in the path
```

## Mapping

```{r}
# 'index' - name has to be the same that was given in the previous command
align(index = "index_subread/chr19_mm10", readfile1 = "/path/to/fastq/file.fastq.gz", nthreads = 8) # "nthreads" - its worth to change it
# on my computer one fastq file mapped approximately 5 min 

# mapping all fastq files from the folder

fastq_files <- list.files(path = "fastq/", pattern = ".fastq.gz$", full.names = T)
head(fastq_files)
align(index = "index_subread/chr19_mm10", readfile1 = fastq_files, nthreads = 12) # its worth to change it
```

## Uniquely mapped reads

```{r}
align(index = "index_subread/chr19_mm10", readfile1 = fastq_files, unique = T, nthreads = 12,
      output_file = paste(fastq_files,"unique","BAM",sep=".")) # "nthreads" - its worth to change it

```

## Mapping quality check

```{r}
bam_files <- list.files(path = "fastq/", pattern = ".unique.BAM$", full.names = T) # "$ at the end of the pattern" - to avoid files with the same name but with further differences, e.g. .unique.BAM.summary
props <- propmapped(files=bam_files)
props
```

# Read counting

```{r}
fc <- featureCounts(bam_files, annot.ext = "ref/Mus_musculus.GRCm39.104.gtf.gz", isGTFAnnotationFile = TRUE, 
                    GTF.featureType = "exon", GTF.attrType = "gene_id")

# check structure of the 'fc'
names(fc)
View(fc$annotation)
```

```{r}
# check dimension
dim(fc$counts)
fc$annotation

# show first rows of the count table

head(fc$counts) # row names - Entrez gene identifiers

counts = as.data.frame(fc$counts) # creates data frame with counts
write.table(counts, file = "mus_counts.txt", sep = "\t", dec = ".") # save data frame with counts
```

```{r}
fc2 <- featureCounts(bam_files, annot.inbuilt = "mm10", 
                    GTF.featureType = "exon", GTF.attrType = "gene_id")

# check structure of the 'fc2'
names(fc2)
fc_diff <- setdiff(fc$annotation, fc2$annotation)
# x = fc2$annotation[grepl('3680155\\b', fc2$annotation$Start),] # just a test
```

# Finish

```{r}
# Session info with package versions
sessionInfo()
```